frontend/src/main/scala/bloop/engine/tasks/Tasks.scala

```diff
@@ -66,21 +66,25 @@ object Tasks {
       )
     }
 
-    val visited = scala.collection.mutable.HashSet.empty[Dag[Project]]
     def compileTree(dag: Dag[Project]): Task[List[CompileResult]] = {
-      if (visited.contains(dag)) Task.now(Nil)
-      else {
-        visited.add(dag)
-        dag match {
-          case Leaf(project) => Task(List(compile(project)))
-          case Parent(project, dependencies) =>
-            val downstream = Task.wanderUnordered(dependencies)(compileTree).map(_.flatten)
-            downstream.flatMap { results =>
-              if (results.exists(_._2 == FailedResult)) Task.now(results)
-              else Task(compile(project)).map(r => r :: results)
-            }
+      val visited = scala.collection.mutable.HashSet.empty[Dag[Project]]
```

Comment:
Why is commit 4cc1ad4c39770be29879a8b2884879078f22371b necessary? Looking at the diff it looks to me like it's only nested one level deeper without changing anything?

Reply:
It's better to scope `visited` in `compileTree` rather than the `compile` function so that it's idempotent. Otherwise adding printlns before the actual use of the function makes return the latter one an empty list.